File: src/0784-insert-into-a-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        """
        In recursion, we return the base case
        And for every if statement we must have a return statement, as without it
        the previous connections to the nodes are ignored
        Basically the return statement is to unravel the recursion tree
        """
        # If I reached the end of the tree wheres theres no leaf nodes
        if not root:
            new_node = TreeNode(val)
            return new_node
        if val < root.val:
            # I need to go left
            root.left = self.insertIntoBST(root.left, val)
            return root
        else:
            root.right = self.insertIntoBST(root.right, val)
            return root

File: src/0784-insert-into-a-binary-search-tree/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function insertIntoBST(root: TreeNode | null, val: number): TreeNode | null {
    // if > than current val insert at right
    // if < than current val insert at left
    // if at the end null, just insert
    if (!root) {
        return new TreeNode(val);
    }
    else if (root.val < val) {
        root.right = insertIntoBST(root.right, val);
    }
    else {
        root.left = insertIntoBST(root.left, val);
    }

    return root;
};
